domain/machine/state/placement.go

```diff
@@ -0,0 +1,559 @@
+		return netNode, []coremachine.Name{
+			machineName,
+		}, errors.Capture(err)
+
+	case deployment.PlacementTypeMachine:
+		// Look up the existing machine by name (example: 0 or 0/lxd/0) and then
+		// return the associated net node UUID.
+		netNodeUUID, err := getMachineNetNodeUUIDFromName(ctx, tx, preparer, coremachine.Name(directive.Directive))
+		return netNodeUUID, nil, errors.Capture(err)
+
+	case deployment.PlacementTypeContainer:
+		// The placement is container scoped (example: lxd or lxd:0). If there
+		// is no directive, we need to create a parent machine (the next in the
+		// sequence) with the associated net node UUID. With a directive we need
+		// to look up the existing machine and place it there. Then we need to
+		// create a child machine for the container and link it to the parent
+		// machine.
+		machineUUID, machineName, err := acquireParentMachineForContainer(ctx, tx, preparer, directive.Directive, platform, nil, clock)
+		if err != nil {
+			return "", nil, errors.Capture(err)
+		}
+
+		// Use the container type to determine the scope of the container.
+		// For example, lxd.
+		scope := directive.Container.String()
+		childNetNode, childMachineName, err := insertChildMachineForContainerPlacement(ctx, tx, preparer, machineUUID, machineName, scope, platform, nonce, clock)
+		if err != nil {
+			return "", nil, errors.Errorf("inserting child machine for container placement: %w", err)
+		}
+		return childNetNode, []coremachine.Name{
+			machineName,
+			childMachineName,
+		}, nil
+
+	case deployment.PlacementTypeProvider:
+		// The placement is handled by the provider, so we need to create a
+		// machine for the net node and then insert the provider placement
+		// for the machine.
+		machineName, err := nextMachineSequence(ctx, tx, preparer)
+		if err != nil {
+			return "", nil, errors.Capture(err)
+		}
+
+		machine, netNode, err := InsertMachineAndNetNode(ctx, tx, preparer, machineName, platform, nonce, clock)
+		if err != nil {
+			return "", nil, errors.Capture(err)
+		}
+		if err := insertMachineProviderPlacement(ctx, tx, preparer, machine, directive.Directive); err != nil {
+			return "", nil, errors.Errorf("inserting machine provider placement: %w", err)
+		}
+		return netNode, []coremachine.Name{
+			machineName,
+		}, nil
+
+	default:
+		return "", nil, errors.Errorf("invalid placement type: %v", directive.Type)
+	}
+}
+
+func nextMachineSequence(ctx context.Context, tx *sqlair.TX, preparer domain.Preparer) (coremachine.Name, error) {
+	namespace := domainmachine.MachineSequenceNamespace
+	seq, err := sequencestate.NextValue(ctx, preparer, tx, namespace)
+	if err != nil {
+		return "", errors.Errorf("getting next machine sequence: %w", err)
+	}
+
+	return coremachine.Name(strconv.FormatUint(seq, 10)), nil
+}
+
+func InsertMachineAndNetNode(
+	ctx context.Context,
+	tx *sqlair.TX,
+	preparer domain.Preparer,
+	machineName coremachine.Name,
+	platform deployment.Platform,
+	nonce *string,
+	clock clock.Clock,
```

Comment:
Pass these as `InsertMachineAndNetNodeArgs`

Reply:
Same, this is part of the follow-up that I already have lined up.